### ISSUE TITLE
build CAGRA index with PQ use cpu

### ISSUE THREAD

**Issue by neezeeyee** [facebookresearch/faiss#4386]
Our scenario involves building an index on the CPU and serving it on the GPU.

I can build an IndexHNSWCAGRA on cpu and copy it to the GPU, but IndexHNSWCAGRA does not support PQ compression. Is there a way to achieve this? It seems that IndexHNSWCAGRA does not inherits from IndexHNSWPQ. lookslike cagra which build on the gpu if we use IVFPQBuildCagra, we can use PQ.

**Comment by mnorris11** [comment]
I don't think it works right now; I tried something like the below and it crashes. 

Recommendation: If your issue is the memory impact is too large, you can try GPU IVFPQ through CUVS. It can support higher M values up to the dimensions of the vectors, as long as M is a factor of d (i.e. M=64 or 128 for d=256)

```
import faiss
from faiss.contrib.datasets import SyntheticDataset

d = 64
hnsw_M = 32
pq_m = 16
pq_nbits = 8
ds: SyntheticDataset = SyntheticDataset(d, 1000, 1000, 100)

index = faiss.IndexHNSWCagra(d, hnsw_M)
index.storage = faiss.IndexPQ(d, pq_m, pq_nbits)
```

(cc @tarang-jain @cjnolet since it relates to CAGRA, maybe I am missing something?)

**Comment by neezeeyee** [comment]
IVFPQ is not as fast as HNSWCAGRA on the gpu. is there any other recommend way to implement?  (will rabitQ support on the gpu?

**Comment by tarang-jain** [comment]
Thats correct. `IndexHNSWCagra` does not have quantization. This is a good idea to allow having the dataset in the quantized format if HNSW will be used for search (but on the other hand, this wont work if an HNSWCagra index is to be converted to CAGRA for search on GPU since CAGRA will require the raw vectors). There is a version of CAGRA (called CAGRA-Q) that uses quantized vectors during search, but that is not a part of Faiss yet. I'll think more on this next week.

**Comment by neezeeyee** [comment]
@tarang-jain  can cagraQ be trained on the cpu and search on the gpu?  Our scenario only allow trains on the cpu . or can u recommend some other algothrim which search faster more than gpu IVFPQ && can be traind on the cpu && quantization

**Comment by tarang-jain** [comment]
`IndexHNSWCagra` has `IndexFlat*` as its storage index, so the snippet that @mnorris11 shared wouldn't work. The reason for this is to be able to supply the full set of raw vectors to CAGRA when converting to a CAGRA index (and vice versa i.e. copy the full set of vectors to CPU when converting a CAGRA index to HNSW since CAGRA does not expose any quantized form of the base vectors). Conversion from an HNSW index that has quantized vectors in its storage to CAGRA is not supported. Furthermore, the form of quantization in CAGRA-Q is VQ (Vector Quantization), which is a little different from PQ in that VQ only uses one codebook for all the subquantizers whereas PQ uses a separate one for each subquantizer. Therefore, there is no direct conversion from product quantized vectors to CAGRA-Q's VQ data store that can be supported at the moment.

@neezeeyee I think the quickest way to support this would be to have two types of data stores in `IndexHNSWCagra` -- one for storing the raw float vectors and the other would be the regular `storage` that could be quantized as per the user's choice. The important piece is to contain the raw vectors inside the HNSW struct. So when converting an HNSW index to a CAGRA index, we can copy those raw vectors. The vice versa, that is converting CAGRA to HNSW would make the `storage` a flat storage. So when such an IndexHNSWCagra index, that is copied from a CAGRA index, is created, the search would use the `IndexFlat*` storage.

**Comment by tarang-jain** [comment]
I can create a PR for this idea in the next few days.

**Comment by mnorris11** [comment]
> I can create a PR for this idea in the next few days.

Sounds good; I'll keep this issue open until then. (Perhaps we can add this issue as an explanation for context in the PR description.)
